File: app-analytics-hub-android/library/src/main/java/com/amazon/appanalyticshub/EventFactory.java
/*
 * Created by thontn on 8/22/18.
 * Copyright © 2018 Amazon.com, Inc. or its affiliates. All rights reserved.
 */
package com.amazon.appanalyticshub;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


/**
 * <p>
 * Event Factory is used to create multiple events from same source and event type.
 * </p>
 */
public class EventFactory {

    private String source;
    private String eventType;

    public EventFactory() {
    }

    /**
     * Creates an Event Factory object for the given a source and event type.
     *
     * @param source    of all the event being created by this factory
     * @param eventType of all the event being created by this factory
     */
    public EventFactory(@Nullable String source, @Nullable String eventType) {
        this.source = source;
        this.eventType = eventType;
    }

    /**
     * Creates an Event Factory object for the given a source and event type.
     *
     * @param source    of all the event being created by this factory
     * @param eventType of all the event being created by this factory
     */
    public EventFactory(@Nullable String source, @NonNull EventType eventType) {
        this(source, eventType.name());
    }

    /**
     * Method to create a new Event with the provided name, source, priority and event type.
     * For creating an event, event type is mandatory so either provide the event
     * type in factory constructor or event factory event data.
     *
     * @param name      - mandatory field for event creation
     * @param source    - optional field for event creation
     * @param priority  - optional field for event creation
     * @param eventType - mandatory field for event creation
     * @return event
     */
    public Event createEvent(@NonNull String name,
                             @Nullable String source,
                             @Nullable Priority priority,
                             @NonNull EventType eventType) {
        return this.createEvent(name, source, priority, eventType.name());
    }

    /**
     * Method to create a new Event with the provided name, source, priority and event type.
     * For creating an event, event type is mandatory so either provide the event
     * type in factory constructor or event factory event data.
     *
     * @param name      - mandatory field for event creation
     * @param source    - optional field for event creation
     * @param priority  - optional field for event creation
     * @param eventType - mandatory field for event creation. Either event factory should have a
     *                  default event type or you have to provide it in this method.
     * @return event
     */
    public Event createEvent(@NonNull String name,
                             @Nullable String source,
                             @Nullable Priority priority,
                             @Nullable String eventType) {
        source = !TextUtils.isEmpty(source) ? source : this.source;
        eventType = !TextUtils.isEmpty(eventType) ? eventType : this.eventType;
        if (TextUtils.isEmpty(eventType)) {
            eventType = Constants.UNKNOWN_TYPE;
        }
        return new Event(name, source, eventType, priority);
    }

    /**
     * Method to create a new Event with the provided name.
     * For creating an event, event type is mandatory so either provide the event
     * type in factory constructor or event factory event data.
     *
     * @param name - mandatory field for event creation
     * @return event
     */
    public Event createEvent(@NonNull String name) {
        return this.createEvent(name, source, null, eventType);
    }

    /**
     * Method to create a new Event with the provided name and priority.
     * For creating an event, event type is mandatory so either provide the event
     * type in factory constructor or event factory event data.
     *
     * @param name     - mandatory field for event creation
     * @param priority - optional field for event creation
     * @return event
     */
    public Event createEvent(@NonNull String name,
                             @Nullable Priority priority) {
        return this.createEvent(name, source, priority, eventType);
    }
}

File: app-analytics-hub-react-native/src/core/event-factory.ts
import { EventType, UNKNOWN_TYPE } from './constants';
import { Event } from './event';
import { IEventFactory, IEventFactoryEventData } from './types';

export class EventFactory implements IEventFactory {
  /**
   * Event type for the events created using this factory. this value can be
   * override by sending a different value while creating the event.
   */
  private eventType?: string | EventType;

  /**
   * Source for the all events created using this factory
   */
  private source?: string;

  /**
   * Creates an Event Factory object for a source.
   * @param options Options to fix in generated events.
   */
  constructor({
    source,
    eventType
  }: {
    /**
     * The source to use for events created by this factory.
     */
    source?: string;
    /**
     * The event type to use for events created by this factory.
     */
    eventType?: string | EventType;
  }) {
    this.source = source;
    this.eventType = eventType!;
  }

  /**
   * Method to create a new Event with the provided source and data.
   * For creating an event, event type is mandatory so either provide the event
   * type in factory constructor or event factory event data.
   *
   * @param {IEventFactoryEventData} eventData
   * @returns {IEvent} would return undefined if event type is not
   * present else would return a new Event instance.
   */
  createEvent(eventData: IEventFactoryEventData): Event {
    let eventType = eventData.eventType ? eventData.eventType : this.eventType;
    if (!eventType) {
      if (__DEV__) {
        throw new Error('No eventType specified.');
      } else {
        eventType = UNKNOWN_TYPE;
      }
    }
    const source = eventData.source ? eventData.source : this.source;
    return new Event(eventData.name, eventType, source, eventData.priority);
  }
}
